### PR TITLE
Add Error for Duplicate New G Suite Users

### PR DESCRIPTION
**Pull request by cbauerman**
#### Changes proposed in this Pull Request

* Add Error for Duplicate New G Suite Users in new user list
Screen Shot 2019-06-24 at 3 44 20 PM

Note: Because each new user component controls when to validate, the error will appear in each component with the same mailbox as you navigate/ give input to it. This quirk is ok because the correct error will appear 

#### Testing instructions
1. Navigate to `/devdocs/design/g-suite-examples`
2. Add two users, give both the same mailbox
3. Confirm that the error message appears on both

### CODE REVIEW

**Comment by sirreal** [comment]
A drive-by unsolicited existential question: Did we implement validation for this form? Did we consider leveraging a library?

We have redux-form sitting around in Calypso, although that lib is a bit heavy and seems to be falling out of favor. There's also [Formik](https://github.com/jaredpalmer/formik) and [Final Form](https://github.com/final-form/final-form) (seems to be redux-form successor, same author) worth keeping an eye on.

There's likely benefit to standardizing on a single, shared validation library.

cc: @jsnajdr

**Comment by jsnajdr** [comment]
@blowery told me the other day that @simison has been experimenting with Formik, but that it didn't work very nicely with our Calypso form components?

File: client/lib/gsuite/new-users.ts [hunk 1]
```diff
@@ -86,10 +85,38 @@ const validateOverallEmailAgainstExistingEmails = (
...
+	value: mailBox,
+	error:
+		mailboxesByCount[ mailBox ] > 1
+			? i18n.translate( 'Please use a unique mailbox for each user.' )
+			: previousError,
+} );
+
+const validateNewUsersAreUnique = ( users: GSuiteNewUser[] ) => {
+	const mailboxesByCount: { [mailbox: string]: number } = countBy(
+		users.map( ( { mailBox: { value: mailBox } } ) => mailBox )
```
Comment (jsnajdr) [inline client/lib/gsuite/new-users.ts]: `countBy( users, 'mailBox.value' )` should do the same thing and is more concise.

**Comment by sirreal** [comment]
> @blowery told me the other day that @simison has been experimenting with Formik, but that it didn't work very nicely with our Calypso form components?

The only thing I can think of is some experimentation we did with Formik for the simple payments block. We ended up dropping it because it didn't work well with Gutenberg inputs, which didn't provide access to raw change events.

**Comment by jsnajdr** [comment]
> The only thing I can think of is some experimentation we did with Formik for the simple payments block.

Yes, that's it. I'd be really happy if we found a form management library that's a good fit for our needs and used it consistently.

**Comment by cbauerman** [comment]
In this case I replaced one set of bespoke validation functions with another 🤷‍♂ However, since we specifically own this form I would be more than happy to try a lib and figure out how to make it work with our underlying form components in this narrow case and document how it works.

@sirreal @jsnajdr between `Formik` and `Final Form` do we have any sort of preference?

**Comment by simison** [comment]
> > @blowery told me the other day that @simison has been experimenting with Formik, but that it didn't work very nicely with our Calypso form components? 
>
> The only thing I can think of is some experimentation we did with  Formik for the simple payments block. We ended up dropping it because it didn't work well with Gutenberg inputs, which didn't provide access to raw change events.

Yep, we tried Formik and I quite enjoyed the flexible API it offers but hit some bumps with Gutenberg's not-so-flexible form components. I got it working, it just wasn't ideal.

I'd assume we wouldn't have those issues in Calypso and even in the core they're solvable with few PRs.

**Comment by jsnajdr** [comment]
> between Formik and Final Form do we have any sort of preference?

I've only used Redux Form so far. I enjoyed using it, the API is nice and very flexible. The downside, however, is that it uses Redux which is not a great fit here: the form state reacts to every keystroke, focus and blur event etc. and that eventually leads to performance problem if Redux Form becomes ubiquitous in a large app like Calypso.

Also, the form state is almost always local to one particular form and different parts of the app don't share the state. Easy cooperation between different parts of the app is a use case where Redux shines, but again, not a good fit here.

Formik and Final Form both have a good reputation, I've been told that the API is very similar to Redux Form, but I haven't personally used them yet. So I don't have a strong opinion.

If you want and can dedicate the time, trying out one of the libraries in this G Suite form and reporting about your experience would be tremendously valuable: forms are everywhere and if we find a good solution, the chosen library would be an important part of our Automattic Interface Toolkit (see also p3N5Fy-HX-p2).

**Comment by simison** [comment]
Might be worth asking Gutenberg/core folks if they've experimented with these and if they have plans to include some kind of form toolkit in the core. If so, I could very well imagine they'd just wrap one of the existing libs and we should then probably try to pick the one they'd likely pick.

**Comment by jsnajdr** [comment]
> Might be worth asking Gutenberg/core folks if they've experimented with these and if they have plans to include some kind of form toolkit in the core.

The way how your suggestion is phrased means that there would be no input into the process of choosing a library from the Calypso side. I think we should be active in that process, instead of passively waiting for "them" 🙂

**Comment by youknowriad** [comment]
Like we adapted several Calypso components to create Gutenberg components in the beginning, a decision made by Calypso can inform future decisions made by Gutenberg.

When it comes to the validation and form libraries, we didn't have a big use-case so far in Gutenberg so we didn't explore that area yet. I expect that the more we adopt Gutenberg patterns in other WP Admin screens, the more obvious this need will become.

I think there are two aspects in that discussion: 

**UI:** I think we should be open to support validation states in our "input" components (TextControl, SelectControl...) in `@wordpress/components`

**Validation and Form Lib:** That's a big unknown for us as the need is not obvious yet as stated above but we'd happy to learn from you all.

**Comment by mtias** [comment]
Error states in input elements would be good to support in the `wordpress/components` packages.

**Comment by cbauerman** [comment]
@sirreal @jsnajdr It turns out that I will not have the bandwidth to take the time to experiment with form validation here in the next month. 

While I still think this is a good place to pursue the question of investigating a standard way of validation, for this particular PR I think I will ask my team to review & close it.

If I have another opportunity I will open a fresh PR and tag the relevant people.

File: client/lib/gsuite/new-users.ts [hunk 2]
```diff
@@ -86,10 +85,38 @@ const validateOverallEmailAgainstExistingEmails = (
 			: mailBoxError,
 } );
 
+const clearPreviousErrors = ( users: GSuiteNewUser[] ) => {
+	return users.map( user => mapValues( user, field => removePreviousErrors( field ) ) );
+};
+
+const validateNewUserMailboxIsUnique = (
+	{ value: mailBox, error: previousError }: GSuiteNewUserField,
```
Comment (stephanethomas) [inline client/lib/gsuite/new-users.ts]: `mailBox` should probably be written as `mailbox` to be consistent with `mailboxesByCount`.

**Comment by sirreal** [comment]
> @sirreal @jsnajdr It turns out that I will not have the bandwidth to take the time to experiment with form validation here in the next month.

No problem, sorry if I derailed this PR at all that wasn't my intention. Thanks for being open and interested in exploring this space 🙌

**Pull request merged**